File: frontend/src/api/documents.ts
import { apiCore, callApi } from './core';
import { Document, DocumentCreate, IndexStatus } from '../types/api';
import axios from 'axios';

export class DocumentApi {
  // --- Dokument-Endpunkte ---

  async createDocument(data: DocumentCreate): Promise<Document> {
    // DIREKT /api/v1 verwenden, nicht erst /v1
    const url = `/api/v1/documents`;
    
    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'X-API-Key': apiCore.getApiKey() || '',
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(data),
      redirect: 'follow',
      mode: 'same-origin',
      credentials: 'same-origin'
    });
    
    if (!response.ok) {
      throw new Error(`HTTP error! Status: ${response.status}`);
    }
    
    return await response.json() as Document;
  }

  async getDocuments(tenantId: string, limit = 1000, offset = 0): Promise<Document[]> {
    try {
      if (!apiCore.getApiKey()) {
        throw new Error('API-Key ist nicht gesetzt');
      }

      console.log('Starte Dokument-Anfrage mit callApi');
      console.log('Mit API-Key:', apiCore.getApiKey());
      console.log('Für Tenant:', tenantId);
      
      // Der Endpunkt erwartet tenant_id als Query-Parameter
      // DIREKT /api/v1 verwenden, nicht erst /v1
      // WICHTIG: Keine Trailing-Slashes verwenden, die zu Redirects führen könnten
      const url = `/api/v1/documents?tenant_id=${encodeURIComponent(tenantId)}`;
      console.log('Verwende URL:', url);
      
      // Die callApi-Funktion würde normalerweise /api/v1 hinzufügen, was hier doppelt wäre
      // Daher verwenden wir fetch direkt mit spezifischen Optionen, die Redirects erlauben
      const response = await fetch(url, {
        method: 'GET',
        headers: {
          'X-API-Key': apiCore.getApiKey() || '',
          'Content-Type': 'application/json'
        },
        // Weiterleitungen erlauben
        redirect: 'follow',
        // Wichtig: Nur Anfragen an die gleiche Herkunft zulassen
        mode: 'same-origin',
        // Wichtig: Credentials nur für gleiche Herkunft senden
        credentials: 'same-origin'
      });
      
      if (!response.ok) {
        throw new Error(`HTTP error! Status: ${response.status}`);
      }
      
      return await response.json() as Document[];
    } catch (error) {
      console.error("Fehler beim Abrufen der Dokumente:", error);
      throw error;
    }
  }

  async getDocument(tenantId: string, documentId: string): Promise<Document> {
    // DIREKT /api/v1 verwenden, nicht erst /v1
    const url = `/api/v1/documents/${documentId}?tenant_id=${encodeURIComponent(tenantId)}`;
    
    const response = await fetch(url, {
      method: 'GET',
      headers: {
        'X-API-Key': apiCore.getApiKey() || '',
        'Content-Type': 'application/json'
      },
      redirect: 'follow',
      mode: 'same-origin',
      credentials: 'same-origin'
    });
    
    if (!response.ok) {
      throw new Error(`HTTP error! Status: ${response.status}`);
    }
    
    return await response.json() as Document;
  }

  async deleteDocument(tenantId: string, documentId: string): Promise<void> {
    // DIREKT /api/v1 verwenden, nicht erst /v1
    const url = `/api/v1/documents/${documentId}?tenant_id=${encodeURIComponent(tenantId)}`;
    
    const response = await fetch(url, {
      method: 'DELETE',
      headers: {
        'X-API-Key': apiCore.getApiKey() || '',
        'Content-Type': 'application/json'
      },
      redirect: 'follow',
      mode: 'same-origin',
      credentials: 'same-origin'
    });
    
    if (!response.ok) {
      throw new Error(`HTTP error! Status: ${response.status}`);
    }
    
    return;
  }

  async uploadCsv(tenantId: string, file: File, titleColumn = 'title', contentColumn = 'content', sourceColumn?: string): Promise<any> {
    const formData = new FormData();
    formData.append('file', file);
    formData.append('title_column', titleColumn);
    formData.append('content_column', contentColumn);
    formData.append('tenant_id', tenantId);
    if (sourceColumn) {
      formData.append('source_column', sourceColumn);
    }

    // FormData muss mit fetch gesendet werden, da callApi es nicht nativ unterstützt
    const response = await fetch(`/api/v1/documents/upload/csv`, {
      method: 'POST',
      headers: {
        'X-API-Key': apiCore.getApiKey() || '',
      },
      body: formData,
      redirect: 'follow',
      mode: 'same-origin',
      credentials: 'same-origin'
    });

    if (!response.ok) {
      throw new Error(`HTTP error! Status: ${response.status}`);
    }

    return await response.json();
  }

  async uploadJson(tenantId: string, file: File): Promise<any> {
    const formData = new FormData();
    formData.append('file', file);
    formData.append('tenant_id', tenantId);

    // FormData muss mit fetch gesendet werden
    const response = await fetch(`/api/v1/documents/upload/json`, {
      method: 'POST',
      headers: {
        'X-API-Key': apiCore.getApiKey() || '',
      },
      body: formData,
      redirect: 'follow',
      mode: 'same-origin',
      credentials: 'same-origin'
    });

    if (!response.ok) {
      throw new Error(`HTTP error! Status: ${response.status}`);
    }

    return await response.json();
  }

  async uploadMarkdown(tenantId: string, file: File, source?: string): Promise<any> {
    const formData = new FormData();
    formData.append('file', file);
    formData.append('tenant_id', tenantId);
    
    if (source) {
      formData.append('source', source);
    }

    // FormData muss mit fetch gesendet werden
    const response = await fetch(`/api/v1/documents/upload/markdown`, {
      method: 'POST',
      headers: {
        'X-API-Key': apiCore.getApiKey() || '',
      },
      body: formData,
      redirect: 'follow',
      mode: 'same-origin',
      credentials: 'same-origin'
    });

    if (!response.ok) {
      throw new Error(`HTTP error! Status: ${response.status}`);
    }

    return await response.json();
  }

  async uploadPdf(tenantId: string, file: File, source?: string): Promise<any> {
    const formData = new FormData();
    formData.append('file', file);
    formData.append('tenant_id', tenantId);
    
    if (source) {
      formData.append('source', source);
    }

    // FormData muss mit fetch gesendet werden
    const response = await fetch(`/api/v1/documents/upload/pdf`, {
      method: 'POST',
      headers: {
        'X-API-Key': apiCore.getApiKey() || '',
      },
      body: formData,
      redirect: 'follow',
      mode: 'same-origin',
      credentials: 'same-origin'
    });

    if (!response.ok) {
      throw new Error(`HTTP error! Status: ${response.status}`);
    }

    return await response.json();
  }

  // --- Dokument-Status und Indizierung ---

  // Dokument-Status abrufen
  async getDocumentStatus(tenantId: string, documentId: string): Promise<{ status: IndexStatus; error?: string }> {
    console.log(`Retrieving document status for document ${documentId} in tenant ${tenantId}`);
    
    if (!tenantId) {
      console.error('getDocumentStatus: tenantId is required');
      return { status: IndexStatus.NICHT_INDIZIERT, error: 'Tenant-ID fehlt' };
    }
    
    if (!documentId) {
      console.error('getDocumentStatus: documentId is required');
      return { status: IndexStatus.NICHT_INDIZIERT, error: 'Dokument-ID fehlt' };
    }

    try {
      // DIREKT /api/v1 verwenden, nicht erst /v1
      const url = `/api/v1/documents/${documentId}/status?tenant_id=${encodeURIComponent(tenantId)}`;
      
      const response = await fetch(url, {
        method: 'GET',
        headers: {
          'X-API-Key': apiCore.getApiKey() || '',
          'Content-Type': 'application/json'
        },
        redirect: 'follow',
        mode: 'same-origin',
        credentials: 'same-origin'
      });
      
      if (!response.ok) {
        throw new Error(`HTTP error! Status: ${response.status}`);
      }
      
      const data = await response.json() as {status: IndexStatus};

      console.log(`Document status retrieved successfully: ${JSON.stringify(data)}`);
      return { status: data.status };
    } catch (error) {
      console.error('Error retrieving document status:', error);
      return { 
        status: IndexStatus.NICHT_INDIZIERT, 
        error: error instanceof Error ? error.message : 'Unbekannter Fehler' 
      };
    }
  }
  
  // Ein einzelnes Dokument neu indizieren
  async reindexDocument(tenantId: string, documentId: string): Promise<any> {
    console.log(`API: Starte Reindexierung für Dokument ${documentId} von Tenant ${tenantId}`);
    try {
      if (!tenantId || !documentId) {
        console.error('Reindexierung fehlgeschlagen: Tenant-ID oder Dokument-ID fehlt');
        throw new Error('Tenant-ID und Dokument-ID müssen angegeben werden');
      }

      // Zuerst das Dokument abrufen, um die vollständigen Daten zu haben
      const document = await this.getDocument(tenantId, documentId);
      
      // DIREKT /api/v1 verwenden, nicht erst /v1
      const url = `/api/v1/documents/${documentId}/reindex?tenant_id=${encodeURIComponent(tenantId)}`;
      
      const response = await fetch(url, {
        method: 'POST',
        headers: {
          'X-API-Key': apiCore.getApiKey() || '',
          'Content-Type': 'application/json'
        },
        redirect: 'follow',
        mode: 'same-origin',
        credentials: 'same-origin',
        body: JSON.stringify(document) // Dokumentdaten im Request-Body senden
      });
      
      if (!response.ok) {
        throw new Error(`HTTP error! Status: ${response.status}`);
      }
      
      return await response.json();
    } catch (error) {
      console.error('Reindexierung fehlgeschlagen:', error);
      const errorMessage = error instanceof Error ? error.message : String(error);
      throw new Error(`Dokument konnte nicht reindexiert werden: ${errorMessage}`);
    }
  }
  
  // Alle Dokumente eines Tenants neu indizieren
  async reindexAllDocuments(tenantId: string): Promise<any> {
    try {
      // DIREKT /api/v1 verwenden, nicht erst /v1
      const url = `/api/v1/documents/reindex-all?tenant_id=${encodeURIComponent(tenantId)}`;
      
      const response = await fetch(url, {
        method: 'POST',
        headers: {
          'X-API-Key': apiCore.getApiKey() || '',
          'Content-Type': 'application/json'
        },
        redirect: 'follow',
        mode: 'same-origin',
        credentials: 'same-origin'
      });
      
      if (!response.ok) {
        throw new Error(`HTTP error! Status: ${response.status}`);
      }
      
      const result = await response.json();
      
      // Wir prüfen den Status der Indizierung im Hintergrund und geben eine sofortige Antwort
      this.pollBulkIndexingStatus(tenantId);
      
      return {
        success: true,
        message: "Indizierung aller Dokumente gestartet",
        ...result
      };
    } catch (error: any) {
      console.error(`Fehler beim Indizieren aller Dokumente für Tenant ${tenantId}:`, error);
      
      // Bei Serverfehlern noch einmal versuchen
      if (error.status >= 500) {
        console.log("Serverfehler beim Massenindizieren, versuche erneut in 5 Sekunden...");
        
        return new Promise((resolve, reject) => {
          setTimeout(async () => {
            try {
              // DIREKT /api/v1 verwenden, nicht erst /v1
              const url = `/api/v1/documents/reindex-all?tenant_id=${encodeURIComponent(tenantId)}`;
              
              const response = await fetch(url, {
                method: 'POST',
                headers: {
                  'X-API-Key': apiCore.getApiKey() || '',
                  'Content-Type': 'application/json'
                },
                redirect: 'follow',
                mode: 'same-origin',
                credentials: 'same-origin'
              });
              
              if (!response.ok) {
                throw new Error(`HTTP error! Status: ${response.status}`);
              }
              
              const retryResult = await response.json();
              
              console.log("Wiederholungsversuch der Massenindizierung erfolgreich");
              
              // Polling im Hintergrund starten
              this.pollBulkIndexingStatus(tenantId);
              
              resolve({
                success: true,
                message: "Indizierung aller Dokumente gestartet (nach Wiederholung)",
                ...retryResult
              });
            } catch (retryError) {
              console.error("Auch der Wiederholungsversuch der Massenindizierung ist fehlgeschlagen:", retryError);
              reject({
                error: "Massenindizierung fehlgeschlagen",
                details: retryError,
                message: "Die Dokumente konnten nicht indiziert werden. Bitte prüfen Sie die Serverprotokolle."
              });
            }
          }, 5000);
        });
      }
      
      throw {
        error: "Massenindizierung fehlgeschlagen",
        details: error,
        message: error.data?.message || "Ein unbekannter Fehler ist aufgetreten."
      };
    }
  }
  
  // Hintergrund-Polling für Massenindizierung
  private async pollBulkIndexingStatus(tenantId: string, interval = 5000, maxAttempts = 60): Promise<void> {
    let attempts = 0;
    
    const checkStatus = async () => {
      try {
        // DIREKT /api/v1 verwenden, nicht erst /v1
        const url = `/api/v1/documents/indexing-status?tenant_id=${encodeURIComponent(tenantId)}`;
        
        const response = await fetch(url, {
          method: 'GET',
          headers: {
            'X-API-Key': apiCore.getApiKey() || '',
            'Content-Type': 'application/json'
          },
          redirect: 'follow',
          mode: 'same-origin',
          credentials: 'same-origin'
        });
        
        if (!response.ok) {
          console.error(`HTTP error bei Abruf des Indizierungsstatus! Status: ${response.status}`);
          return;
        }
        
        const status = await response.json();
        
        console.log(`Indizierungsstatus für Tenant ${tenantId}:`, status);
        
        // Prüfen, ob die Indizierung abgeschlossen ist
        if (status.completed || attempts >= maxAttempts) {
          console.log("Indizierung abgeschlossen oder maximale Versuche erreicht");
          return;
        }
        
        // Weiter pollen
        attempts++;
        setTimeout(checkStatus, interval);
      } catch (error) {
        console.error("Fehler beim Abrufen des Indizierungsstatus:", error);
        
        // Bei Fehlern weiter versuchen, aber nicht endlos
        if (attempts < maxAttempts) {
          attempts++;
          setTimeout(checkStatus, interval);
        }
      }
    };
    
    // Polling starten
    checkStatus();
  }
  
  async updateDocument(tenantId: string, documentId: string, data: any): Promise<any> {
    // DIREKT /api/v1 verwenden, nicht erst /v1
    const url = `/api/v1/documents/${documentId}?tenant_id=${encodeURIComponent(tenantId)}`;
    
    const response = await fetch(url, {
      method: 'PUT',
      headers: {
        'X-API-Key': apiCore.getApiKey() || '',
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(data),
      redirect: 'follow',
      mode: 'same-origin',
      credentials: 'same-origin'
    });
    
    if (!response.ok) {
      throw new Error(`HTTP error! Status: ${response.status}`);
    }
    
    return await response.json();
  }

  // Weaviate-Status eines Dokuments abrufen (alte Methode)
  async getDocumentWeaviateStatus(tenantId: string, documentId: string): Promise<any> {
    // Delegieren an die neue Methode
    return this.getWeaviateStatus(tenantId, documentId);
  }

  // Weaviate-Status eines Dokuments abrufen
  async getWeaviateStatus(tenantId: string, documentId: string): Promise<any> {
    console.log(`API: Rufe Weaviate-Status für Dokument ${documentId} von Tenant ${tenantId} ab`);
    try {
      if (!tenantId || !documentId) {
        console.error('Weaviate-Status-Abruf fehlgeschlagen: Tenant-ID oder Dokument-ID fehlt');
        throw new Error('Tenant-ID und Dokument-ID müssen angegeben werden');
      }

      // DIREKT /api/v1 verwenden, nicht erst /v1
      const url = `/api/v1/documents/${documentId}/weaviate-status?tenant_id=${encodeURIComponent(tenantId)}`;
      
      try {
        const response = await fetch(url, {
          method: 'GET',
          headers: {
            'X-API-Key': apiCore.getApiKey() || '',
            'Content-Type': 'application/json'
          },
          redirect: 'follow',
          mode: 'same-origin',
          credentials: 'same-origin'
        });
        
        if (!response.ok) {
          // Wenn der Backend-Endpunkt fehlschlägt, liefern wir einen Status zurück, der anzeigt,
          // dass wir den Status nicht abrufen konnten, aber die UI kann weiterlaufen
          console.warn(`Weaviate-Status konnte nicht abgerufen werden: HTTP Status ${response.status}`);
          return {
            status: 'unknown',
            error: `Konnte Status nicht abrufen: HTTP Status ${response.status}`
          };
        }
        
        return await response.json();
      } catch (fetchError) {
        // Bei Netzwerkfehler oder anderen Problemen
        console.warn('Netzwerkfehler beim Abrufen des Weaviate-Status:', fetchError);
        return {
          status: 'unknown',
          error: 'Netzwerkfehler beim Abrufen des Status'
        };
      }
    } catch (error) {
      console.error('Weaviate-Status-Abruf fehlgeschlagen:', error);
      // Fallback-Status zurückgeben, damit die UI nicht abstürzt
      return {
        status: 'unknown',
        error: error instanceof Error ? error.message : 'Unbekannter Fehler beim Abrufen des Status'
      };
    }
  }
}

// Singleton-Instanz
export const documentApi = new DocumentApi(); 

File: backend/app/api/v1/documents.py
from fastapi import APIRouter, Depends, HTTPException, status, UploadFile, File, Form
from typing import List, Optional
from pydantic import parse_obj_as
import json
import pandas as pd
import io
import uuid
from datetime import datetime
from sqlalchemy.orm import Session
from ...db.models import Document, DocumentCreate, DocumentModel
from ...services.weaviate_service import weaviate_service
from ...core.security import get_tenant_id_from_api_key
from ...db.session import get_db
from ...services import document_service
from ...schemas.document import WeaviateStatus
from ...core.auth import get_api_key

router = APIRouter()


@router.post("/", response_model=Document, status_code=status.HTTP_201_CREATED)
async def create_document(
    document: DocumentCreate,
    tenant_id: str = Depends(get_tenant_id_from_api_key),
    db: Session = Depends(get_db)
):
    """Fügt ein neues Dokument zur Wissensbasis hinzu."""
    # Weaviate-Dokument erstellen
    doc_id = weaviate_service.add_document(
        tenant_id=tenant_id,
        title=document.title,
        content=document.content,
        metadata=document.doc_metadata,
        source=document.source
    )
    
    # Datenbank-Dokument erstellen
    db_document = DocumentModel(
        id=doc_id,
        tenant_id=tenant_id,
        title=document.title,
        content=document.content,
        doc_metadata=document.doc_metadata,
        source=document.source
    )
    
    db.add(db_document)
    db.commit()
    db.refresh(db_document)
    
    return Document.model_validate(db_document)


@router.options("/", include_in_schema=False)
@router.options("/{document_id}", include_in_schema=False)
@router.options("/upload/csv", include_in_schema=False)
@router.options("/upload/json", include_in_schema=False)
async def options_documents():
    """Handler für OPTIONS-Anfragen an Documents-Endpunkte."""
    return {}


@router.get("/", response_model=List[Document])
async def get_documents(
    tenant_id: str,
    limit: int = 1000,
    offset: int = 0,
    api_key: str = Depends(get_api_key),
    db: Session = Depends(get_db)
):
    """Ruft alle Dokumente eines Tenants ab."""
    try:
        print(f"Suche Dokumente für Tenant: {tenant_id}")
        # Dokumente aus der Datenbank abrufen
        db_documents = db.query(DocumentModel).filter(
            DocumentModel.tenant_id == tenant_id
        ).offset(offset).limit(limit).all()
        
        print(f"Gefundene Dokumente: {len(db_documents)}")
        for doc in db_documents:
            print(f"Dokument: ID={doc.id}, Titel={doc.title}, Created={doc.created_at}")
        
        result = [Document.model_validate(doc) for doc in db_documents]
        print(f"Serialisierte Dokumente: {len(result)}")
        return result
    except Exception as e:
        print(f"Fehler beim Abrufen der Dokumente: {e}")
        import traceback
        print(traceback.format_exc())
        return []


@router.delete("/{document_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_document(
    document_id: str,
    tenant_id: str = Depends(get_tenant_id_from_api_key),
    db: Session = Depends(get_db)
):
    """Löscht ein Dokument aus der Wissensbasis."""
    # Dokument aus der Datenbank löschen
    db_document = db.query(DocumentModel).filter(
        DocumentModel.id == document_id,
        DocumentModel.tenant_id == tenant_id
    ).first()
    
    if not db_document:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Dokument nicht gefunden"
        )
    
    # Aus Weaviate löschen
    success = weaviate_service.delete_document(
        tenant_id=tenant_id,
        doc_id=document_id
    )
    
    if success:
        db.delete(db_document)
        db.commit()
    
    return None


@router.post("/upload/csv", status_code=status.HTTP_201_CREATED)
async def upload_csv(
    file: UploadFile = File(...),
    title_column: str = Form("title"),
    content_column: str = Form("content"),
    source_column: Optional[str] = Form(None),
    tenant_id: str = Depends(get_tenant_id_from_api_key),
    db: Session = Depends(get_db)
):
    """
    Lädt eine CSV-Datei hoch und fügt die Daten als Dokumente zur Wissensbasis hinzu.
    
    - Die CSV muss mindestens eine Spalte für Titel und Inhalt haben
    - Optional kann eine Spalte für die Quelle angegeben werden
    """
    try:
        # CSV-Datei lesen
        contents = await file.read()
        df = pd.read_csv(io.BytesIO(contents))
        
        # Prüfen, ob die erforderlichen Spalten vorhanden sind
        if title_column not in df.columns or content_column not in df.columns:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"CSV muss die Spalten '{title_column}' und '{content_column}' enthalten"
            )
        
        # Dokumente hinzufügen
        added_docs = []
        for _, row in df.iterrows():
            title = row[title_column]
            content = row[content_column]
            source = row.get(source_column) if source_column and source_column in df.columns else None
            
            # Metadaten aus allen anderen Spalten erstellen
            metadata = {}
            for col in df.columns:
                if col not in [title_column, content_column, source_column]:
                    metadata[col] = row[col]
            
            # Dokument in Weaviate hinzufügen
            doc_id = weaviate_service.add_document(
                tenant_id=tenant_id,
                title=title,
                content=content,
                metadata=metadata,
                source=source
            )
            
            # Dokument in der Datenbank speichern
            db_document = DocumentModel(
                id=doc_id,
                tenant_id=tenant_id,
                title=title,
                content=content,
                doc_metadata=metadata,
                source=source
            )
            db.add(db_document)
            
            added_docs.append({
                "id": doc_id,
                "title": title
            })
        
        # Alle Dokumente in der Datenbank speichern
        db.commit()
        
        return {
            "message": f"{len(added_docs)} Dokumente erfolgreich hinzugefügt",
            "documents": added_docs
        }
    
    except Exception as e:
        db.rollback()  # Rollback bei Fehlern
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Fehler beim Verarbeiten der CSV-Datei: {str(e)}"
        )


@router.post("/upload/json", status_code=status.HTTP_201_CREATED)
async def upload_json(
    file: UploadFile = File(...),
    tenant_id: str = Depends(get_tenant_id_from_api_key),
    db: Session = Depends(get_db)
):
    """
    Lädt eine JSON-Datei hoch und fügt die Daten als Dokumente zur Wissensbasis hinzu.
    
    Die JSON-Datei sollte ein Array von Objekten mit mindestens "title" und "content" enthalten.
    Optional können "source" und "metadata" angegeben werden.
    """
    try:
        # JSON-Datei lesen
        contents = await file.read()
        data = json.loads(contents)
        
        if not isinstance(data, list):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="JSON muss ein Array von Objekten sein"
            )
        
        # Dokumente hinzufügen
        added_docs = []
        for item in data:
            # Als DocumentCreate parsen
            try:
                doc = parse_obj_as(DocumentCreate, item)
            except Exception:
                # Überprüfen, ob mindestens Titel und Inhalt vorhanden sind
                if "title" not in item or "content" not in item:
                    continue
                
                # Manuell ein Dokument erstellen
                doc = DocumentCreate(
                    title=item["title"],
                    content=item["content"],
                    source=item.get("source"),
                    doc_metadata=item.get("metadata", {})
                )
            
            # Dokument in Weaviate hinzufügen
            doc_id = weaviate_service.add_document(
                tenant_id=tenant_id,
                title=doc.title,
                content=doc.content,
                metadata=doc.doc_metadata,
                source=doc.source
            )
            
            # Dokument in der Datenbank speichern
            db_document = DocumentModel(
                id=doc_id,
                tenant_id=tenant_id,
                title=doc.title,
                content=doc.content,
                doc_metadata=doc.doc_metadata,
                source=doc.source
            )
            db.add(db_document)
            
            added_docs.append({
                "id": doc_id,
                "title": doc.title
            })
        
        # Alle Dokumente in der Datenbank speichern
        db.commit()
        
        return {
            "message": f"{len(added_docs)} Dokumente erfolgreich hinzugefügt",
            "documents": added_docs
        }
    
    except Exception as e:
        db.rollback()  # Rollback bei Fehlern
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Fehler beim Verarbeiten der JSON-Datei: {str(e)}"
        )


@router.post("/upload/markdown", status_code=status.HTTP_201_CREATED)
async def upload_markdown(
    file: UploadFile = File(...),
    source: Optional[str] = Form(None),
    tenant_id: str = Depends(get_tenant_id_from_api_key),
    db: Session = Depends(get_db)
):
    """
    Lädt eine Markdown-Datei hoch und fügt sie als Dokument zur Wissensbasis hinzu.
    Der Dateiname wird als Titel verwendet, der Inhalt als Content.
    """
    try:
        # Markdown-Datei lesen
        contents = await file.read()
        content = contents.decode('utf-8')
        
        # Titel aus Dateinamen extrahieren (ohne .md-Endung)
        title = file.filename
        if title.lower().endswith('.md'):
            title = title[:-3]
        
        # Dokument in Weaviate hinzufügen
        doc_id = weaviate_service.add_document(
            tenant_id=tenant_id,
            title=title,
            content=content,
            metadata={"file_type": "markdown"},
            source=source
        )
        
        # Dokument in der Datenbank speichern
        db_document = DocumentModel(
            id=doc_id,
            tenant_id=tenant_id,
            title=title,
            content=content,
            doc_metadata={"file_type": "markdown"},
            source=source
        )
        db.add(db_document)
        db.commit()
        db.refresh(db_document)
        
        return {
            "message": "Markdown-Datei erfolgreich hinzugefügt",
            "document": {
                "id": doc_id,
                "title": title
            }
        }
    
    except Exception as e:
        db.rollback()  # Rollback bei Fehlern
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Fehler beim Verarbeiten der Markdown-Datei: {str(e)}"
        )


@router.get("/{document_id}/weaviate-status", response_model=WeaviateStatus)
async def get_document_weaviate_status(
    document_id: str,
    tenant_id: str = Depends(get_tenant_id_from_api_key),
    db: Session = Depends(get_db)
):
    """Weaviate-Status eines Dokuments abrufen"""
    document = db.query(DocumentModel).filter(
        DocumentModel.id == document_id,
        DocumentModel.tenant_id == tenant_id
    ).first()
    
    if not document:
        raise HTTPException(status_code=404, detail="Dokument nicht gefunden")
    
    status = await weaviate_service.get_document_status(tenant_id, document_id)
    return status


@router.post("/{document_id}/reindex")
async def reindex_document(
    document_id: str,
    tenant_id: str = Depends(get_tenant_id_from_api_key),
    db: Session = Depends(get_db)
):
    """Ein einzelnes Dokument neu indizieren"""
    document = db.query(DocumentModel).filter(
        DocumentModel.id == document_id,
        DocumentModel.tenant_id == tenant_id
    ).first()
    
    if not document:
        raise HTTPException(status_code=404, detail="Dokument nicht gefunden")
    
    try:
        await weaviate_service.reindex_document(document)
        return {"message": "Dokument erfolgreich neu indiziert"}
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Fehler beim Neu-Indizieren des Dokuments: {str(e)}"
        )


@router.post("/reindex-all")
async def reindex_all_documents(
    tenant_id: str = Depends(get_tenant_id_from_api_key),
    db: Session = Depends(get_db)
):
    """Alle Dokumente neu indizieren"""
    try:
        # Alle Dokumente des Tenants abrufen
        documents = db.query(DocumentModel).filter(
            DocumentModel.tenant_id == tenant_id
        ).all()
        
        if not documents:
            return {"message": "Keine Dokumente zum Neu-Indizieren gefunden"}
        
        # Dokumente in Document-Schema konvertieren
        doc_schemas = [Document.model_validate(doc) for doc in documents]
        
        # Alle Dokumente neu indizieren
        await weaviate_service.reindex_all_documents(tenant_id, doc_schemas)
        
        return {"message": f"{len(documents)} Dokumente erfolgreich neu indiziert"}
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Fehler beim Neu-Indizieren der Dokumente: {str(e)}"
        )


@router.get("/{document_id}/status", response_model=WeaviateStatus)
async def get_document_status(
    document_id: str,
    tenant_id: str = Depends(get_tenant_id_from_api_key),
    db: Session = Depends(get_db)
):
    """Ruft den Indizierungsstatus eines Dokuments ab."""
    # Prüfen, ob das Dokument existiert
    db_document = db.query(DocumentModel).filter(
        DocumentModel.id == document_id,
        DocumentModel.tenant_id == tenant_id
    ).first()
    
    if not db_document:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Dokument nicht gefunden"
        )
    
    # Status von Weaviate abrufen
    status = await weaviate_service.get_document_status(tenant_id, document_id)
    return status 